controller/src/scheduler.rs

```diff
@@ -0,0 +1,160 @@
+            MaintenanceTime::Oneshot(s) => sleep_until_next_scheduled_time(s).await,
+        }
+    }
+
+    /// Determine when controller needs discontinue updates.
+    /// specific trigger time => never discontinue updates.
+    /// maintenance window (time window): discontinue updates when current is outside of a scheduled window.
+    pub fn discontinue_updates(&self) -> bool {
+        match self {
+            MaintenanceTime::Windows(s) => {
+                if s.includes(Utc::now()) {
+                    return false;
+                }
+            }
+            MaintenanceTime::Oneshot(s) => return false,
+        }
+        true
+    }
+}
+
+async fn sleep_until_next_scheduled_time(schedule: &Schedule) -> Result<()> {
+    let next_scheduled_time = schedule
+        .upcoming(Utc)
+        .next()
+        .context(scheduler_error::GetScheduledDatetimeSnafu)?;
+    let duration_until_next_schedule = next_scheduled_time - Utc::now();
+
+    // covert chrono duration to std duration
+    let duration_until_next_schedule_std = duration_until_next_schedule
+        .to_std()
+        .context(scheduler_error::ConvertToStdDurationSnafu)?;
+
+    event!(
+        Level::INFO,
+        next_schedule_time = ?next_scheduled_time,
+        sleep = duration_until_next_schedule_std.as_secs(),
+        "Began sleeping until next scheduled time point."
+    );
+    tokio::time::sleep(duration_until_next_schedule_std).await;
+    Ok(())
+}
+
+fn calculate_duration_between_scheduled_datetimes(
+    brupop_schedule: &Schedule,
+) -> Result<chrono::Duration> {
+    let next_schedule_datetime = brupop_schedule
+        .upcoming(Utc)
+        .next()
+        .context(scheduler_error::GetScheduledDatetimeSnafu)?;
+
+    let next_next_schedule_datetime = brupop_schedule
+        .upcoming(Utc)
+        .nth(1)
+        .context(scheduler_error::GetScheduledDatetimeSnafu)?;
+
+    let duration_between_each_schedule_datetime =
+        next_next_schedule_datetime - next_schedule_datetime;
+
+    Ok(duration_between_each_schedule_datetime)
+}
+
+pub mod scheduler_error {
+    use snafu::Snafu;
+
+    #[derive(Debug, Snafu)]
+    #[snafu(visibility(pub))]
+    pub enum Error {
+        #[snafu(display("Unable convert to Std duration due to {}", source))]
+        ConvertToStdDuration { source: chrono::OutOfRangeError },
+
+        #[snafu(display(
+            "Unable to get environment variable '{}' due to : '{}'",
+            variable,
+            source
+        ))]
+        MissingEnvVariable {
+            source: std::env::VarError,
+            variable: String,
+        },
+
+        #[snafu(display("Failed to generate corn expression '{}' due to `{}`", cron, source))]
+        GenerateScheduleFailed {
+            cron: String,
+            source: cron::error::Error,
+        },
+
+        #[snafu(display("Unable to get cron expression schedule scheduled datetime"))]
+        GetScheduledDatetime {},
+    }
+}
```

Comment:
You did a nice job of providing an abstraction to improve the code, but I think it could be a bit better. Please look at what I've done here and notice how I've made almost all of the logic testable.

~~Also note, I'm not sure you're supposed to sleep a controller by using a thread sleep, I thought you were supposed to tell the controller when it should run it's reconciliation loop again.~~ @jpmcb says this is fine.

Also note, technically a cron expression that has a 1 second window in it could be mistaken for a oneshot. I'm not sure if this edge case matters much.

```suggestion
use chrono::{DateTime, Duration, Utc};
use cron::Schedule;
use snafu::{OptionExt, ResultExt};
use std::env;
use std::str::FromStr;
use tracing::{event, Level};

// Defines the cron expression scheduler related env variable names
const SCHEDULER_CRON_EXPRESSION_ENV_VAR: &str = "SCHEDULER_CRON_EXPRESSION";

/// The module-wide result type.
type Result<T> = std::result::Result<T, scheduler_error::Error>;

pub struct BrupopCronScheduler {
    schedule: cron::Schedule,
    schedule_type: ScheduleType,
}

impl BrupopCronScheduler {
    pub fn from_environment() -> Result<Self> {
        let chron_expression = env::var(SCHEDULER_CRON_EXPRESSION_ENV_VAR).context(
            scheduler_error::MissingEnvVariableSnafu {
                variable: SCHEDULER_CRON_EXPRESSION_ENV_VAR.to_string(),
            },
        )?;
        Self::from_string(chron_expression)
    }

    /// Determine when controller needs discontinue updates.
    /// specific trigger time => never discontinue updates.
    /// maintenance window (time window): discontinue updates when current is outside of a scheduled window.
    pub fn discontinue_updates(&self) -> bool {
        self.discontinue_updates_impl(Utc::now())
    }

    pub async fn sleep_until_next(&self) -> Result<()> {
        let now = Utc::now();
        let d = self.duration_to_next(now)?;
        let std_duration = std_duration(&d)?;
        let seconds = std_duration.as_secs();
        let next_schedule_time = now + d;
        event!(
            Level::INFO,
            next_schedule_time = ?next_schedule_time,
            sleep = seconds,
            "Sleeping until next scheduled time point."
        );
        // TODO - i don't think this is how you are supposed to sleep a k8s controller, aren't you
        // supposed to tell it when to its reconciliation loop again?
        tokio::time::sleep(std_duration).await;
        Ok(())
    }

    // TODO - use this function in tests
    fn from_string<S: AsRef<str>>(cron_expression: S) -> Result<Self> {
        let schedule = Schedule::from_str(cron_expression.as_ref()).context(
            scheduler_error::GenerateScheduleFailedSnafu {
                cron: cron_expression.as_ref(),
            },
        )?;
        let schedule_type = determine_schedule_type(&schedule)?;
        Ok(BrupopCronScheduler {
            schedule,
            schedule_type,
        })
    }

    // TODO - write tests for this function
    fn duration_to_next(&self, now: DateTime<Utc>) -> Result<Duration> {
        let next_scheduled_time = self
            .schedule
            .after(&now)
            .next()
            .context(scheduler_error::GetScheduledDatetimeSnafu)?;
        Ok(next_scheduled_time - now)
    }

    // TODO - write tests for this function
    fn discontinue_updates_impl(&self, now: DateTime<Utc>) -> bool {
        match self.schedule_type {
            ScheduleType::Windowed => {
                if self.schedule.includes(now) {
                    return false;
                }
            }
            ScheduleType::Oneshot => return false,
        }
        true
    }
}

// Cron expression can be configured to a time window or a specific trigger time.
// specific trigger time: 0 0 10 * * Mon */Every Monday at 10AM.
// maintenance window (time window): * * 10-12 * * MON */Every Monday between 10:00 and 12:00.

// brupop controller needs to use different logics to deal with specific trigger time or
// maintenance window (time window)
// => specific trigger time: trigger brupop update and complete all waitingForUpdate nodes.
// => maintenance window (time window): trigger brupop update within time window. If current
// time isn't within the time window, controller shouldn't have any action on it.
enum ScheduleType {
    Windowed,
    Oneshot,
}

// TODO - write test of this function
fn duration_between_next_two_points(schedule: &Schedule, from: DateTime<Utc>) -> Result<Duration> {
    let first_time = schedule
        .after(&from)
        .next()
        .context(scheduler_error::GetScheduledDatetimeSnafu)?;

    let second_time = schedule
        .after(&from)
        .nth(1)
        .context(scheduler_error::GetScheduledDatetimeSnafu)?;

    Ok(second_time - first_time)
}

// TODO - this isn't exactly correct, what if my schedule has a 1-second window in it?
fn determine_schedule_type(schedule: &Schedule) -> Result<ScheduleType> {
    let duration_between_each_schedule_datetime =
        duration_between_next_two_points(&schedule, Utc::now())?;

    Ok(
        if duration_between_each_schedule_datetime.num_seconds() == 1 {
            ScheduleType::Windowed
        } else {
            ScheduleType::Oneshot
        },
    )
}

fn std_duration(d: &Duration) -> Result<std::time::Duration> {
    Ok(d.to_std()
        .context(scheduler_error::ConvertToStdDurationSnafu)?)
}

pub mod scheduler_error {
    use snafu::Snafu;

    #[derive(Debug, Snafu)]
    #[snafu(visibility(pub))]
    pub enum Error {
        #[snafu(display("Unable convert to Std duration due to {}", source))]
        ConvertToStdDuration { source: chrono::OutOfRangeError },

        #[snafu(display(
            "Unable to get environment variable '{}' due to : '{}'",
            variable,
            source
        ))]
        MissingEnvVariable {
            source: std::env::VarError,
            variable: String,
        },

        #[snafu(display("Failed to generate corn expression '{}' due to `{}`", cron, source))]
        GenerateScheduleFailed {
            cron: String,
            source: cron::error::Error,
        },

        #[snafu(display("Unable to get cron expression schedule scheduled datetime"))]
        GetScheduledDatetime {},
    }
}
```

Reply:
> Also note, technically a cron expression that has a 1 second window in it could be mistaken for a oneshot. I'm not sure if this edge case matters much.

Yes, users can set oneshot cron expression which has a 1 second window in it like `1,2,3,4 10 10 * * * *`, which means trigger oneshot scheduler on everyday `10:10:1`, `10:10:2`,  `10:10:3`, `10:10:4`. They have 1 sec gap between each of them, and the scheduler will recognize them as time window.
However, first I think this situation is meaningless edge case. I'm not sure what is the purpose of that. Meanwhile,  `1,2,3,4 10 10 * * * *` is actually same as `1-4 10 10 * * * *`, so it makes sense that the scheduler recognize it as time window.